adr/0026-schema.md

```diff
@@ -0,0 +1,615 @@
+In summary, a `wait` block will be auto migrated if, and only if:
+  - it is `onDeploy.after`
+  - the condition is `ready`, `available` or equivalent
+  - The kind is not a custom resource.
+- There may be further, more holistic changes to how actions works within Zarf that would have a significant affect on the schema. This will be covered in a separate ADR.
+
+### BDD scenarios
+The following are (behavior driven development)[https://en.wikipedia.org/wiki/Behavior-driven_development] scenarios to provide context of what Zarf will do in specific situations given the above decisions.
+
+#### v1 create with deprecated keys
+- *Given* Zarf version is v1
+- *and* the `zarf.yaml` has no apiVersion or deprecated keys
+- *when* the user runs `zarf package create`
+- *then* they will receive an error and be told to run `zarf dev update-schema` or how to migrate off cosign key paths or how to use flavors over groups depending on the error
+
+#### v0 create -> v1 deploy
+- *Given*: A package is created with Zarf v0
+- *and* that package has deprecated keys that can be automatically migrated (required, scripts, & set variables)
+- *when* the package is deployed with Zarf v1
+- *then* the keys will be automatically migrated & the package will be deployed without error.
+
+#### v0 create with removed feature -> v1 deploy
+- *Given*: A package is created with Zarf v0
+- *and* that package has deprecated keys that cannot be automatically migrated (groups, cosignKeyPath)
+- *when* the package is deployed with Zarf v1
+- *then* then deploy of that package will fail and the user will be instructed to update their package
+
+#### v1 create -> v0 deploy
+- *Given*: A package is created with Zarf v1
+- *and* that package uses keys that don't exist in v0
+- *when* the package is deployed with Zarf v0
+- *then* Zarf v0 will deploy the package without issues. If there are fields unrecognized by the v0 schema, then the user will be warned they are deploying a package that has features that do not exist in the current version of Zarf.
+
+## Consequences
+- As long as the only deprecated features in a package have migration path, and the package was built after the feature was deprecated so migrations were run, Zarf will be successful both creating a package with v1 and deploying with v0, and creating a package with v0 and deploying with v1.
+- Users of deprecated group, cosignKeyPath, and action.Wait outside of onDeploy might be frustrated if their packages, created v0, error out on Zarf v1, however this is preferable to unexpected behavior occurring in the cluster.
+- Users may be frustrated that they have to run `zarf dev update-schema` to edit their `zarf.yaml` to remove the deprecated fields and add `apiVersion`.
+- We will have to have two different schema types which will be mostly be duplicate code. However the original type should never change, which mitigates much of the issue.
```

Comment:
```suggestion
- We will have to have two different schema types which will be mostly duplicated. However the original type should never change, which mitigates much of the issue.
```

In this case do we mean versions of schemas or types, as in data-types, of values?

Reply:
I meant schema types, but that sentence was confusing. Clarified the sentence and added the examples we based this idea on, hopefully that helps